### PR TITLE
Allow case sensitive renaming

### PR DESCRIPTION
**Pull request by justjack2347**
## Description

This PR allows for the renaming of files with different capitalisation.

## Changes

Removed two instances of making the filenames lowercase when comparing (in file-rename.ts and fsal-directory.ts).

## Additional information

The check is still done by lower casing the file names before file creation since I believe that removing this could cause problems on case-insensitive file systems.

This pull request addresses issue #4322 

Tested on: Windows 11
University of Adelaide - Software Process Improvement
Additional Contributors:

- @josh-133
- @Mr-Verb

### CODE REVIEW

**Comment by nathanlesage** [comment]
Thanks for this PR, I appreciate your support.

The first change in the command it looks proper – case-sensitive renaming is possible, but when the app checks for duplicates, it does so case-insensitive to ensure cross-platform compatibility.

However, this fine-grained check is missing from the second file, the `renameChild` utility function. There, it fails to check if there is already a file case-insensitively, thus allowing case-sensitive renaming _but also overwriting of existing files_.

Would you mind adding the necessary checks for that as well? That would be awesome.

**Comment by justjack2347** [comment]
Hi @nathanlesage, thank you for your comments. I have had a look and play around and completely understand your concern but cannot find an example of where we can overwrite existing files and so if you had an instance in mind, that would be great, but otherwise I will have more of a think. I have tried using the test-gui to rename a file to another filename with different capitalisation and it raises an override warning. The override warning is what we are looking for isn't it?

If the `renameChild` utility function renames the file/directory and then rebuilds the directory tree (from my understanding), and is called after the rename file or rename directory functions, would these functions include the protections for file overwriting. I am likely wrong here so please let me know.

Thanks again

**Comment by nathanlesage** [comment]
I think the issue is that there are two functions responsible for renaming – the corresponding command as well as `renameChild`. Since `renameChild` may be called programmatically without the command running, we have to ensure that both functions employ the appropriate safeguards. I think the confusion stems from the fact that right now there is kind of a shared responsibility, and we may want to move all the logic to the `renameChild` function (i.e. that the command basically just executes the function).

What the `renameChild` function thus should check is:

0. Is the new name exactly the same as the old name? Then just silently return, as nothing has to change (and by not touching the file, we prevent an array of other event listeners to be active)
1. Is the new name equal to the old name regardless of capitalization? In that case, allow the rename.
2. Next, if it is not, is there any other file that already has the new name (regardless of capitalization)? If so, ask before overwriting.
3. Next, if the name is new and unique, just perform the renaming.

What the command should check is:

* Is the file modified?

This check needs to stay in the command as the `renameChild` function has no access to the service container and therefore cannot check if there are currently unsaved changes to it.

Does this make sense?

**Comment by justjack2347** [comment]
Yes, it did make sense, thank you. So I have added some of the additional checks in. The only check that cannot be performed within the `renameChild` function is the overwriting check, since that requires access to the service container of the app and which it does not have access to. To counter this, I did add a parameter (which is set to false as default) called `allowOverwrite.` In the case where the new name is not equal to the old name case insensitively, it will check if the function is allowed to overwrite files (set by parameter). While it would be preferable to have the functionality for asking about an overwrite in the function, this way of implementation would mean that if this function is called, it will default to not automatically overwriting. 

It is worth noting that setting the `allowOverwrite` boolean isn't required for the current implementation since in the rename-file code, it runs the overwrite check and if that we allow the overwrite, it just deletes the file currently there.

Also, regarding moving the logic to the `renameChild` function, I think that most of the code in the rename-file code is dealing with the `appServiceContainer` and so I don't believe this can be easily transferred across.

Is there any concerns about this?

**Comment by justjack2347** [comment]
Hi @nathanlesage, just wondering whether there were any thoughts about this? Is defaulting to safe behaviour (not overwriting) an acceptable approach to ensure using the function will not result in the files being overwritten unless the `allowOverwrite` is explicitly set to true?

**Comment by nathanlesage** [comment]
Hey, I took a Short break from coding, but slowly getting back to it. I need to take another look at your new changes, then I'll reply more thoroughly again!

File: source/app/service-providers/fsal/fsal-directory.ts
Comment (nathanlesage) [inline source/app/service-providers/fsal/fsal-directory.ts:405]: I would suggest renaming this parameter to `forceOverwrite`. The reason is that, somehow (this may just be my personal preference, so feel free to comment on this suggestion by me), `allowOverwrite` sounds to me as if there was user choice involved, whereas `forceOverwrite` has a more programmatic sound to it.

Since we now do have the proper checks in the file renaming command, I don't think we ever need to ask the user to set this parameter. Instead, it makes sense to include this parameter purely for potential future use if we at some point allow Zettlr to programmatically rename files. Since it works properly in this implementation, I'd suggest we definitely keep it for future use, and if the parameter is never used, then, well, so be it.

**Review by nathanlesage — APPROVED** [review]
I think this all looks very solid. I had one comment, but that doesn't involve change if you disagree. Please either implement my suggestion, if you agree it makes sense, or simply comment that you disagree, and afterwards I will merge this PR!

Thanks already for your work here!

File: source/app/service-providers/fsal/fsal-directory.ts
Comment (justjack2347) [reply source/app/service-providers/fsal/fsal-directory.ts]: I think that this change makes sense, change should be part of the pull request now.

Comment (nathanlesage) [reply source/app/service-providers/fsal/fsal-directory.ts]: Wonderful, thank you!

**Comment by nathanlesage** [comment]
Thank you so much for your work! I'll wait until the unit tests have finished, and then merge!

**Pull request merged**
